### ISSUE TITLE
Vector Embedding Search Index

### ISSUE THREAD

**Issue by sjpritchard** [vaticle/typedb#6911]
Please replace every line in curly brackets { like this } with appropriate answers, and remove this line.

## Problem to Solve

I would like to be able to use vector embeddings to perform semantic searching over a TypeDB database, to support use cases like Retrieval Augmented Generation and integration with frameworks such as LlamaIndex. Vector embeddings are now increasingly supported across various database platforms, including Neo4j: https://neo4j.com/docs/cypher-manual/current/indexes-for-vector-search/

## Current Workaround

None

## Proposed Solution

Solution similar to Neo4j: https://neo4j.com/docs/cypher-manual/current/indexes-for-vector-search/

## Additional Information

None

**Comment by flyingsilverfin** [comment]
Thank you for the request and the link to neo's resources! The approach is surely interesting and we should consider it - we haven't built much in the direction of advanced indexing (via Lucene for example) but it is something under consideration. Probably for our next roadmap it's not in scope, but when we develop the following one we can re-visit it.

**Comment by sjpritchard** [comment]
I think that this is a big missed opportunity for TypeDB. Graph databases are having a big resurgence due to AI and Retrieval Augmented Generation, and TypeDB is nowhere to be seen in the ecosystem. See https://python.langchain.com/docs/use_cases/graph/ and https://docs.llamaindex.ai/en/stable/examples/index_structs/knowledge_graph/KnowledgeGraphDemo.html#

**Comment by haikalpribadi** [comment]
Indeed it is something important for us to bring into TypeDB, @sjpritchard ! We'll look into adding it to our roadmap in 2024.

**Comment by sjpritchard** [comment]
![image](https://github.com/user-attachments/assets/4f1c2b5a-4d77-4d86-afa1-337e2774825c)

**Comment by flyingsilverfin** [comment]
Primary things to remember when implementing vector indexing from my background week of work...

# Reference systems

Systems we can reference:
1. Lucene
2. Hannoy (works over KV store, which is interesting)
3. Kuzu (simplified 2-layer sampled graph)

Key takeaway: this area is RIPE for low-level algorithmic, data structure, and data layout/cpu cache optimizations. There is tons of room for SIMD optimization and vectorization, loop unrolls, etc.

HNSW is quite easy to implement - Claude can do it well! The systems with the best alignment with us do:
- build the index in-memory in a large batch, at commit time.
  - keeping the vectors and the graph on-disk or in a complex structure is SLOW and not cache efficient
  - we can transform the vectors from the snapshot buffer into an optimized data structure to build the index
- add unindexed flat segments that are searched in brute force, and compacted into a new index in the background

# Usage

I expect that we'll encourage people using `attribute pdf-embedding @independent @index(hnsw), value vector[1024];` to Bulk Load attributes in large transactions (say, 1-10 million vectors), which will explicitly consume lots of ram and be a slow, long running write transaction operation.

We can even say that we build a new index for 'large' writes to the vector attribute, say if it's > 100k entries. Less than 100k brute forcing the segment is ok.

Querying:

1. we can one function that does everything like, `similarity_search()`
This lets us do the following:
```
match
  let $pdf-embedding in similarity_search(pdf-embedding, [target vector], [ index=true|false ], [ max_similarity = double], [top_k = integer ])
```

This is a weird function with signature `similarity_search(type: Type, vector: vector[_], index? : bool, max_similarity?: double, top_k?: integer)`

Note the first argument is sorta like a generic (?).

And returns an iterator. This is better than having a distance operator because it's controllable (index yes/no, how far to stream answers from, threshold it if you want, etc.

HNSW is also nice because you can iteratively broaden the search and keep pulling answers from the frontier till the index runs out of answers. Note, this will not be all answers.

2. We can do hybrid search my either pulling from the iterator and continuing with the traversal, start with a filtered search and collect a set of vectors to do brute force search over.

**Comment by matthiasautrata** [comment]
Maybe old but some interesting thoughts about what can be done with HNSW: https://docs.cozodb.org/en/latest/releases/v0.6.html

**Comment by flyingsilverfin** [comment]
I just benchmarked Cozo's hnsw build and it's extremely slow for building the hnsw since it routes everything through rocksdb. interesting!

**Comment by matthiasautrata** [comment]
> I just benchmarked Cozo's hnsw build and it's extremely slow for building the hnsw since it routes everything through rocksdb. interesting!

I thought the interesting part in Cozo's implementation was that they didn't wrap the HNSW but distance/similarity became accessible almost like an edge. My mental image: Fuzzy edges or edges with strength.

**Comment by flyingsilverfin** [comment]
I appreciate the input @matthiasautrata ! I think for our purposes, since I'm leaning towards vector indices living on the attribute type, and we don't currently have attribute to attribute edges in TypeDB, it'll probably be a separate index-based jump that integrates via a function call.

**Comment by flyingsilverfin** [comment]
The proposed function-call based approach only supports post-filtering.

However, with some clever query planning and execution, we can also support pre-filtering [ note: this breaks assumptions we have in function execution right now, where the output of a function never reverses flow to affect the execution of the function, so this is non-trivial! ]

```
match           
    let $emb in similarity_search(pdf-embedding, $target);
    $doc isa document, has pdf-embedding $emb, has category "science";
```

Here, if the planner identifies that the space of `$emb` is very small by searching `$doc` with `science` categories first, then producing `$emb` will be a small set and we can do the similary search as a post-filter. 

We can now see that `top_k` is a nonsense parameter: the order of the execution would change the results. For example - f we do the similarity search first, then we get 10 answers, none of which may be from the science category. If we get science docs first, we will probably get 10 documents.

Instead, we should just always use `limit N` to get the top N results! If this needs to be isolated from the rest of the query, users can put it in a function.

```
match 
  let $emb in ...
limit 10;
```

Now, we have one universal way of saying "find me the first (top) 10 things" and that works regardless of query planning order.

**Comment by k4otix** [comment]
(Note - written with AI because your team and Fable/Opus understand this much better than I do!)

---

Hi all — I've been following this issue while building a research project on TypeDB 3.x where similarity search will be the retrieval entry point. Nice to see the groundwork landing in typedb/typeql#456 and typedb-behaviour#429. Since the value-type work is underway, I'll skip to two things from a source-level study I did against `master` (`77fc34bf4`) that affect the *search* phase — easier to accommodate now than to retrofit.

## 1. What does `limit` mean right after `similarity_search`?

The sketch earlier in this thread ends with `similarity_search(...) ... limit 10;` — right call, but it hides a semantics decision. Today `limit` is pure count truncation: without a `sort`, you get the first N in *plan* order, which is an artifact, not a guarantee (the archived spec makes the same point about `offset`). So "top 10 by similarity" either makes the search stage's emission order semantically meaningful, or users write `sort $score; limit 10;` — and `sort` currently materializes its whole input, which against a lazy index stream is a lot of work for ten rows.

The good news, from reading the planner: sortedness is *already* a cost-based physical property inside `match` — sorted-iterator intersections, cost-chosen sort variables, "sort-by data" driving merge decisions. It just never crosses into the pipeline-stage layer. So the ask isn't a new concept, it's a boundary extension: let a stage-level `sort` notice its input is already ordered by the same comparator and become a streaming no-op. The elision is clean — `sort` is stable with input-order tie-break, so skipping it over an already-ordered stream is exactly the identity, and `insert` drains its input regardless of any sort, so write pipelines don't care. Scope it to engine-known producers (a user-declarable "trust me, it's ordered" annotation would be a soundness hole) and keep it cost-based, and it stays an optimization rather than a planner constraint.

Why now: `limit` already composes lazily (it stops pulling the moment its count is reached, including across the inlined function-call boundary), so ordered stream → elided sort → `limit` is the fully lazy "top N" this thread's `limit`-over-`top_k` conclusion points at — with defined semantics instead of plan-order luck.

Smaller, related — updating my note above, since the `given` stage landed in 3.12.0 (#7812): `given` already provides the forward-flow road to pre-filtering. Candidates bound via `given` enter the pipeline as inputs, so restricting similarity to a candidate set never needs information to flow backward into a search. Once the vector value type and a distance function land, the whole pre-filtered path is just composition: `given` the candidates, match their embeddings, score each by distance to the query, `sort`, `limit` — and because a pre-filtered set is small, the materializing `sort` is the right plan there; the laziness concern in this note only bites when the stream is the whole index. If the indexed `similarity_search()` later wants to accept a candidate restriction, `given`-bound inputs offer that with pure forward flow too — no new query constructs required.

## 2. Attribute identity for vectors

Attributes dedup by value, and bit-exact equality is weak identity for embeddings — the same text through the same model can differ in low bits, and `±0.0`/distinct `NaN`s already hash apart for `Double`. Worth an explicit decision rather than an inherited one: near-identical embedding → new attribute? (Hash-of-bytes probably is the right answer, and quantized storage makes it cleaner — the canonical stored bytes *are* the identity.) And presumably vectors are non-keyable like `Double`/`Struct`. Small corner, painful to change once data exists.

---

Thanks for the discussion! I look forward to these features in a future release

_edited: updated the pre-filtering note for the `given` stage that shipped in 3.12.0_
